hax-lib-protocol-macros/src/lib.rs

```diff
@@ -0,0 +1,300 @@
+use quote::quote;
+use syn::{parse, parse_macro_input};
+
+/// This macro takes an `fn` as the basis of an `InitialState` implementation for the state type that is returned by the `fn` (on success).
+///
+/// The `fn` is expected to build the state type from a `Vec<u8>`.
```

Comment:
Giving a slice should work as well, I thought it might be more flexible if the newly created state had ownership over its initial input.

Reply:
Let's add it like this. But we probably want to change this at some point.

The issue is that this requires an allocator so we couldn't use it on embedded things without one. It could for example be interesting to throw this onto [edhoc](https://github.com/openwsn-berkeley/lakers/blob/main/lib/src/edhoc.rs) since they're using hax already.

Reply:
That makes sense! I thought it was okay from an allocation point-of-view since `Vec::new()` doesn't allocate, but hadn't considered `no_std` environments.